src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java

```diff
@@ -501,4 +519,303 @@ public int hashCode() {
+	 */
+	public OGCConcreteGeometryCollection flatten() {
+		if (isFlattened()) {
+			return this;
+		}
+		
+		OGCMultiPoint multiPoint = null;
+		ArrayList<Geometry> polygons = null;
+		OGCMultiLineString polyline = null;
+		GeometryCursor gc = getEsriGeometryCursor();
+		for (Geometry g = gc.next(); g != null; g = gc.next()) {
+			if (g.isEmpty())
+				continue;
+			
+			Geometry.Type t = g.getType();
+
+			if (t == Geometry.Type.Point) {
+				if (multiPoint == null) {
+					multiPoint = new OGCMultiPoint(esriSR);
+				}
+
+				((MultiPoint)multiPoint.getEsriGeometry()).add((Point)g);
+				continue;
+			}
+
+			if (t == Geometry.Type.MultiPoint) {
+				if (multiPoint == null)
+					multiPoint = new OGCMultiPoint(esriSR);
+				
+				((MultiPoint)multiPoint.getEsriGeometry()).add((MultiPoint)g, 0, -1);
+				continue;
+			}
+			
+			if (t == Geometry.Type.Polyline) {
+				if (polyline == null)
+					polyline = new OGCMultiLineString(esriSR);
+				
+				((MultiPath)polyline.getEsriGeometry()).add((Polyline)g, false);
+				continue;
+			}
+			
+			if (t == Geometry.Type.Polygon) {
+				if (polygons == null)
+					polygons = new ArrayList<Geometry>();
+				
+				polygons.add(g);
+				continue;
+			}
+			
+			throw new GeometryException("internal error");//what else?
+		}
+
+		List<OGCGeometry> list = new ArrayList<OGCGeometry>();
+
+		if (multiPoint != null)
+			list.add(multiPoint);
+
+		if (polyline != null)
+			list.add(polyline);
+		
+		if (polygons != null) {
+			GeometryCursor unionedPolygons = OperatorUnion.local().execute(new SimpleGeometryCursor(polygons), esriSR, null);
+			Geometry g = unionedPolygons.next();
+			if (!g.isEmpty()) {
+				list.add(new OGCMultiPolygon((Polygon)g, esriSR));
+			}
+
+		}
+
+		return new OGCConcreteGeometryCollection(list, esriSR);
+	}
+	
+	/**
+	 * Fixes topological overlaps in the GeometryCollecion.
+	 * This is equivalent to union of the geometry collection elements.
+	 * 
+	 * @return A geometry collection that is flattened and has no overlapping elements.
+	 */
+	public OGCConcreteGeometryCollection flattenAndRemoveOverlaps() {
+		ArrayList<Geometry> geoms = new ArrayList<Geometry>();
+		
+		//flatten and crack/cluster
+		GeometryCursor cursor = OGCStructureInternal.prepare_for_ops_(flatten().getEsriGeometryCursor(), esriSR);
+		for (Geometry g = cursor.next(); g != null; g = cursor.next()) {
+			geoms.add(g);
+		}
+		
+		//make sure geometries don't overlap
+		return removeOverlapsHelper_(geoms);
+	}
+	
+	private OGCConcreteGeometryCollection removeOverlapsHelper_(List<Geometry> geoms) {
+		ArrayList<Geometry> result = new ArrayList<Geometry>();
+		for (int i = 0; i < geoms.size(); ++i) {
+			Geometry current = geoms.get(i);
+			if (current.isEmpty())
+				continue;
+			
+			for (int j = i + 1; j < geoms.size(); ++j) {
+				Geometry subG = geoms.get(j);
+				current = OperatorDifference.local().execute(current, subG, esriSR, null);
+				if (current.isEmpty())
```

Comment:
Why breaking on the first geometry that doesn't overlap `current`? Shouldn't we check for all geometries first?

Reply:
We break because the current - subG resulted in empty geometry, so current is completely inside of subG.